packages/perseus-core/src/widgets/matcher/matcher-util.ts

```diff
@@ -0,0 +1,61 @@
+import {CoreUtil} from "@khanacademy/perseus-core";
+
+import type {PerseusMatcherWidgetOptions} from "@khanacademy/perseus-core";
+
+type MatcherInfo = {
+    left: ReadonlyArray<string>;
+    right: ReadonlyArray<string>;
+    orderMatters: boolean;
+    problemNum: number | null | undefined;
+};
+
+export const matcherShuffle = (
+    props: MatcherInfo,
+): {left: ReadonlyArray<string>; right: ReadonlyArray<string>} => {
+    // Use the same random() function to shuffle both columns sequentially
+    const rng = CoreUtil.seededRNG(props.problemNum as number);
+
+    let left;
+    if (!props.orderMatters) {
+        // If the order doesn't matter, don't shuffle the left column
+        left = props.left;
+    } else {
+        left = CoreUtil.shuffle(props.left, rng, /* ensurePermuted */ true);
+    }
+
+    const right = CoreUtil.shuffle(props.right, rng, /* ensurePermuted */ true);
```

Comment:
This `ensurePermuted = true` actually leaks information about the correct answer. We are revealing that the correct answer is _not_ the permutation sent to the client. A cheater could answer the question with better than chance accuracy by choosing some other permutation at random.

Fun fact: a variant of this type of information leak [allowed the Allies to break the Enigma cipher in World War II](https://en.wikipedia.org/wiki/Cryptanalysis_of_the_Enigma)!

> A major weakness of the system, however, was that no letter could be enciphered to itself.

Reply:
Hmm, I see that this line was copied from the component. I think we should keep it as-is to avoid changing the UX.

Reply:
Agree. Cryptographically it's leaking information, but I think we don't want to ever render the correct answer as the starting point.